### PR TITLE
Falcon DSA

### PR DESCRIPTION
**Pull request by Al-Kindi-0**
## Describe your changes
This PR proposes an adaptation of the reference C implementation so that it uses RPO instead of SHAKE256 for the hash-to-point algorithm.

## Checklist before requesting a review
- Repo forked and branch created from `next` according to naming convention.
- Commit messages and codestyle follow [conventions](./CONTRIBUTING.md).
- Relevant issues are linked in the PR description.
- Tests added for new functionality.
- Documentation/comments updated according to changes.

### CODE REVIEW

**Review by bobbinth — COMMENTED** [review]
Not a full review yet, but I left a couple of comments about Rust structs.

In general, I'm wondering if we should organize Rust structs a bit differently to fit the use case of Miden VM a bit better.

First, in our case, the public key will be usually just a commitment to the public key represented by 4 field elements (i.e., `Word`). This is how we'll store it in account storage, for example. Full public key will be provided via advice provider (i.e., `adv_pipe` instructions). So, maybe public key struct should look something like this:
```Rust
PublicKey(Word)
```

The above means that to verify signature, we should be sending the actual public key together with the signature. This will increase size of the signature somewhat, but I don't think this makes a big difference for us. So, the signature struct may look something like this:
```Rust
Signature {
    sig: [u8; SIG_LEN],
    pk: [u8; PK_LEN],
}
```

An alternative is to use signature in key-recovery mode (section 3.12 in Falcon spec paper) - but maybe this will complicate things too much.

Also, the message we'd usually want to sign will be reduced to `Word` before we get to signing part (i.e., we'd compress the message using RPO within the VM before we invoke Falcon signature procedure). Not sure if this affects anything as we can always covert RPO digest into 32 bytes - but thought I'd mentioned nevertheless.

File: src/dsa/falcon/structs/pk.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
+use super::sig::Signature;
+use crate::dsa::falcon::{binder::*, param::*, Polynomial};
+use libc::c_void;
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct PublicKey(pub(crate) [u8; PK_LEN]);
+
+impl PublicKey {
+    // Verify using the reference C implementation
+    pub fn verify_c(&self, message: &[u8], signature: &Signature) -> bool {
```
[the remark this answers is not in the document]
Comment (Al-Kindi-0) [reply src/dsa/falcon/structs/pk.rs:10]: I think we can remove it altogether.

File: src/dsa/falcon/structs/kp.rs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+use rand::Rng;// TODO: substitute with OsRng
+use zeroize::Zeroize;
+
+use super::{PublicKey, SecretKey};
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct KeyPair {
+    pub public_key: PublicKey,
+    pub secret_key: SecretKey,
+}
```
[the remark this answers is not in the document]
Comment (Al-Kindi-0) [reply src/dsa/falcon/structs/kp.rs:12]: Makes sense!

**Comment by Al-Kindi-0** [comment]
I have implemented the suggestions above. The implementation is in #179 and the discussion should be continued there.

**Pull request closed without merging**